### ISSUE TITLE
3.1 – Implement `HybridDirectionalModel`

### ISSUE THREAD

**Issue by P7AC1D** [P7AC1D/WebForum#1]
Excellent — here’s the fully expanded **GitHub-ready Markdown** version of
**Task 3.1 – Implement `HybridDirectionalModel`**, written in the same professional format as your other tasks.

It explains the model’s architecture, purpose, and design rationale clearly — ideal for inclusion in your GitHub backlog or technical documentation.

---

# 3.1 – Implement `HybridDirectionalModel`

### Description

Implement the **Hybrid Directional Model**, a neural architecture that combines an **LSTM encoder** (to capture local temporal dependencies) with a **Transformer encoder** (to model global attention and contextual relationships).

The model’s goal is to predict **short-term directional bias** or **expected N-bar normalized return** given recent market features and, eventually, the current regime context.

This hybrid design blends sequence memory and contextual attention, enabling better generalization to shifting market structures — particularly useful for adaptive trading intelligence systems like **AurumMind**.

---

### Background

Traditional LSTMs capture local sequences effectively but struggle with long-range dependencies.
Transformers, on the other hand, model global relationships using **multi-head self-attention** but lack built-in temporal recurrence.

By stacking them:

* The **LSTM** first encodes localized time patterns (e.g., 30-bar microtrends).
* The **Transformer** then learns cross-temporal relationships and contextual dependencies among features.
* The final **dense head** produces a scalar directional signal (expected return or probability of upward movement).

This combination offers a strong inductive bias for financial time series that exhibit both short-term autocorrelation and long-term contextual dynamics.

---

### Implementation Goals

1. **Model Architecture**

   * Implement in PyTorch as `HybridDirectionalModel`.
   * Architecture flow:

     ```
     Features [T×F]
        ↓
     LSTM Encoder (temporal memory)
        ↓
     Linear Projection to Transformer Dim
        ↓
     Transformer Encoder (contextual attention)
        ↓
     Mean or Last-Token Pooling
        ↓
     Dense Head → Output Scalar
     ```

2. **Configurable Hyperparameters**

   * `feature_dim`, `hidden_dim`, `d_model`, `n_heads`, `num_layers`, `ff_dim`, `dropout`
   * Allow configuration via a YAML/JSON config or constructor arguments.

3. **Output**

   * Scalar output representing either:

     * Expected normalized return (regression), or
     * Probability of upward movement (classification)

4. **Testing**

   * Add a simple unit test to confirm the forward pass works with sample data.
   * Validate tensor dimensions and absence of shape mismatches.

---

### Acceptance Criteria

* Model implemented in PyTorch as `HybridDirectionalModel`
* Architecture includes: **LSTM → projection → Transformer → pooling → dense head**
* Configurable hyperparameters for dimensions, heads, and dropout
* Unit test confirms correct output shape and functional forward pass

---

### Checklist

* [ ] Define model class in `directional_agent.py`
* [ ] Implement LSTM + Transformer encoder architecture
* [ ] Add pooling layer and dense output head
* [ ] Write forward method with clear tensor flow
* [ ] Add unit test verifying output dimensions
* [ ] Save model skeleton and configuration defaults

---

### Example (Skeleton Code)

```python
import torch
import torch.nn as nn

class HybridDirectionalModel(nn.Module):
    def __init__(self, feature_dim=30, hidden_dim=128, d_model=128,
                 n_heads=4, num_layers=2, ff_dim=256, dropout=0.1):
        super().__init__()
        self.lstm = nn.LSTM(feature_dim, hidden_dim, batch_first=True)
        self.proj = nn.Linear(hidden_dim, d_model)
        enc_layer = nn.TransformerEncoderLayer(
            d_model, n_heads, ff_dim, dropout, batch_first=True)
        self.encoder = nn.TransformerEncoder(enc_layer, num_layers)
        self.pool = nn.AdaptiveAvgPool1d(1)
        self.head = nn.Sequential(
            nn.LayerNorm(d_model),
            nn.Linear(d_model, 64),
            nn.ReLU(),
            nn.Linear(64, 1)
        )

    def forward(self, x):
        lstm_out, _ = self.lstm(x)              # [B, T, H]
        x = self.proj(lstm_out)                 # [B, T, D]
        x = self.encoder(x)                     # [B, T, D]
        pooled = x.mean(dim=1)                  # temporal pooling
        return self.head(pooled).squeeze(-1)    # [B]
```

---

### Notes

* This hybrid architecture bridges **temporal memory** and **contextual attention**, giving it an edge in learning both short-term and long-term relationships.
* The output scalar will later feed into the **Policy Agent** as directional context.
* Keep model deterministic and seed-controlled for reproducibility during training.

---

Would you like me to generate the next task (3.2 – Integrate Regime Conditioning into Model Inputs) in the same expanded style so you can keep all of Phase 3 consistent?
